### DISCUSSION TITLE
Is It Possible to Install Python 3.11.3 in Particular with Homebrew?

### DISCUSSION THREAD

**Discussion by id88viper88id** [Homebrew/discussions#5014] [Everyday usage]
### Output of `brew config`

```shell
hubert@Huberts-Mac-Mini ~ % brew config
HOMEBREW_VERSION: 4.2.1
ORIGIN: https://github.com/Homebrew/brew
HEAD: 7817518992920b8dbd15e48a5e6c5ad768f4ba8b
Last commit: 7 days ago
Core tap JSON: 31 Dec 15:43 UTC
Core cask tap JSON: 31 Dec 15:43 UTC
HOMEBREW_PREFIX: /opt/homebrew
HOMEBREW_CASK_OPTS: []
HOMEBREW_MAKE_JOBS: 12
Homebrew Ruby: 3.1.4 => /opt/homebrew/Library/Homebrew/vendor/portable-ruby/3.1.4/bin/ruby
CPU: dodeca-core 64-bit arm_blizzard_avalanche
Clang: 15.0.0 build 1500
Git: 2.43.0 => /opt/homebrew/bin/git
Curl: 8.4.0 => /usr/bin/curl
macOS: 14.2.1-arm64
CLT: 15.1.0.0.1.1700200546
Xcode: N/A
Rosetta 2: false
hubert@Huberts-Mac-Mini ~ %
```

### Output of `brew doctor`

```shell
hubert@Huberts-Mac-Mini ~ % brew doctor
Please note that these warnings are just used to help the Homebrew maintainers
with debugging if you file an issue. If everything you use Homebrew for is
working fine: please don't worry or file an issue; just ignore this. Thanks!

Warning: Some installed formulae are deprecated or disabled.
You should find replacements for the following formulae:
  openssl@1.1

Warning: You have unlinked kegs in your Cellar.
Leaving kegs unlinked can lead to build-trouble and cause formulae that depend on
those kegs to fail to run properly once built. Run `brew link` on these:
  python-jinja
  python-certifi
  python-sympy
  python-mpmath
  python-filelock
  pytorch
  python-networkx
  python-markupsafe
hubert@Huberts-Mac-Mini ~ %
```

### Description of issue

Hello to everyone as this is my very first post on the project :) Pardon me if this is a silly question to ask, but is it possible to install Python 3.11.3 in particular using Homebrew?

The reason I'm asking about this is because I need Python 3.11.3 in particular to upgrade some other Github project to a different version of it. I installed Python 3.10 and Python 3.11 running `brew install python@3.10` and `brew install python@3.11` respectfully, and when I go to the `/opt/homebrew/Cellar/python@3.11` directory, the only Python version there is Python 3.11.6.1 (I don‘t know if that `1` at the end means something like 'Revision 1' or not).

When I run `python --version` alone, the output I get is this:

`zsh: command not found: python`

And when I run `python3 --version`, I get the following:

`Python 3.9.6`. I'll add here that switching the interpreter from ZSH to, say, Bash, does not change anything, as, one way or the other, in the first case both interpreters say that the command is not found, while in the latter case they both indicate the same Python 3.9.6 version.

I infer from the above that Python 3.9.6 is the default version of Python for macOS, which comes pre-installed with the latest version of macOS Sonora, which is 14.2.1 (23C71).

I would like to make it clear that I do not want to override the default version of Python for macOS. What I want to do is to install Python 3.11.3 as a separate version of Python on the system.

When I try to run `brew install python@3.11.3`, I get the following:

```
hubert@Huberts-Mac-Mini ~ % brew install python@3.11.3
==> Downloading https://formulae.brew.sh/api/formula.jws.json
######################################################################### 100.0%
==> Downloading https://formulae.brew.sh/api/cask.jws.json
######################################################################### 100.0%
Warning: No available formula with the name "python@3.11.3". Did you mean python@3.11, python@3.10 or python@3.12?
==> Searching for similarly named formulae and casks...
==> Formulae
python@3.11 ✔              python@3.10 ✔              python@3.12 ✔

To install python@3.11 ✔, run:
  brew install python@3.11 ✔
hubert@Huberts-Mac-Mini ~ %
```

As `Warning: No available formula with the name "python@3.11.3"` suggests, Python 3.11.3 is not available as a formula through Homebrew. The above seems to be the case because when you look at the versions of Python available here:

https://formulae.brew.sh/formula/python@3.11 ,

the generic version available is Python 3.11 (Python 3.11.6 being the currently available stable version, just as the one installed through Homebrew running `brew install python@3.11`), while the other versions available are (in a descending order):

Python 3.12.1
Python 3.10.13
Python 3.9.18
Python 3.8.18
Python 3.7.17

I know that it could take lots of space in general if one were to keep each and every version of every package on the server, but does the above list mean, in fact, that it is impossible to install Python 3.11.3 through Homebrew?

I know I could install Python 3.11.3 using an installer for macOS from https://www.python.org, but then it wouldn‘t be manageable, and perhaps not even executable, as I assume, through Homebrew. Managing and executing packages through Homebrew is more convenient, though, and that is how I would prefer it to be

**Comment by Bo98** [comment]
Not entirely sure why you need 3.11.3 in particular given 3.11.x versions are generally compatible but `pyenv` likely is what you want to install arbitrary versions. `pyenv` is available via Homebrew.

**Comment by id88viper88id** [comment]
Pardon me if I violate any potential rules here by mentioning another GitHub project, but I‘m doing so for the sake of explanation:

There is a package called yt-dlp (an improved, fork version of deprecated youtube-dl) for downloading videos, but also audio only, from YouTube and some other sites. Its argument to download only the indicated videos from playlists doesn‘t work the why it should, and I wanted to report it as a bug, but the creators say that before one does so, they should upgrade yt-dlp from the stable version available to either the so-called 'nightly' version or 'master' version. When earlier I had the `python` command aliased to `/opt/homebrew/Cellar/python@3.11` in `~/.zshrc` and wanted to update yt-dlp to the 'nightly' version, it would tell me that Python 3.11.3 was missing, and that the update couldn‘t be done because of that. I assumed that it was because under `/opt/homebrew/Cellar/python@3.11` there is only the `3.11.6_1` directory, and no `3.11.3` directory anywhere. That made me assume that since there wasn‘t any `3.11.3` directory anywhere, it must be why it said it was missing, and so I removed the aliasing from `~/.zshrc` to later be able to point to anywhere it may be necessary once Python 3.11.3 was installed. This is where we get to the point where I was having trouble installing Python 3.11.3, and that is why I asked about that.
I now see, though, that all this reasoning was wrong… What I did now, as simply taking a shot, was that I replaced `python` from the command for updating yt-dlp from the stable to that so-called 'nightly' version for `python3`, and this time it simply worked…

Reading this and that, it is simply confusing whether one should stick to how certain commands were initially given and use `python` alone, or maybe `python` with the digit of its generic version available and updated to at the time… Well, it was my bad for not trying to slightly change the initial command given. Anyway, thank you for your help :)

I could simply close the topic now, as everything has got cleared out, but I'll leave it open in case anyone wants to add anything :)
